Src/ILGPU/IndexTypes.tt

```diff
@@ -0,0 +1,716 @@
+        /// <returns>The computed linear index of this index.</returns>
+<# if (def.IsIntIndex) { #>
+        public readonly int ComputeLinearIndex(<#= name #> dimension) =>
+            ComputeLinearIndex(dimension, this);
+<# } else { #>
+        readonly int ILinearizableIndex<<#= name #>>.ComputeLinearIndex(
+            <#= name #> dimension)
+        {
+            long linearIndex = ComputeLinearIndex(dimension, this);
+            IndexTypeExtensions.AssertIntIndexRange(linearIndex);
+            return (int)linearIndex;
+        }
+
+        /// <summary>
+        /// Computes the linear index of this index by using the provided n-D dimension.
+        /// </summary>
+        /// <param name="dimension">The dimension for index computation.</param>
+        /// <returns>The computed linear index of this index.</returns>
+        public readonly long ComputeLinearIndex(<#= name #> dimension) =>
+            ComputeLongLinearIndex(dimension);
+<# } #>
+
+        /// <summary>
+        /// Computes the linear index of this index by using the provided n-D dimension.
+        /// </summary>
+        /// <param name="dimension">The dimension for index computation.</param>
+        /// <returns>The computed linear index of this index.</returns>
+        public readonly long ComputeLongLinearIndex(<#= name #> dimension) =>
+            ComputeLinearIndex(dimension, this);
+
+        /// <summary>
+        /// Reconstructs an index from a linear index.
+        /// </summary>
+        /// <param name="linearIndex">The linear index.</param>
+        /// <returns>The reconstructed index.</returns>
+        public readonly <#= name #> ReconstructIndex(int linearIndex) =>
+<# if (def.IsIntIndex) { #>
+            ReconstructIndex(linearIndex, this);
+<# } else { #>
+            ReconstructIndex((long)linearIndex);
+<# } #>
+
+        /// <summary>
+        /// Reconstructs an index from a linear index.
+        /// </summary>
+        /// <param name="linearIndex">The linear index.</param>
+        /// <returns>The reconstructed index.</returns>
+        public readonly <#= name #> ReconstructIndex(long linearIndex)
+<# if (def.IsIntIndex) { #>
+        {
+            IndexTypeExtensions.AssertIntIndexRange(linearIndex);
+            return ReconstructIndex((int)linearIndex, this);
+        }
+<# } else { #>
+        => ReconstructIndex(linearIndex, this);
+<# } #>
+
+        #endregion
+
+        #region IGenericIndex
+
+        /// <summary>
+        /// Returns true if the current index is greater than or equal to 0 and
+        /// is less than the given dimension.
+        /// </summary>
+        /// <param name="dimension">The dimension bounds.</param>
+        /// <returns>True if the current index is inside the given bounds.</returns>
+        public readonly bool InBounds(<#= name #> dimension) =>
+            <#= def.Expression(" & ", p => $"{p} >= 0 & {p} < dimension.{p}") #>;
```

Comment:
I chose `&/|| in favor of `&&/|| to avoid short-circuit evaluation on GPUs in favor of more efficient predicated control flow. These patterns are converted by the `IfConversion` transformation anyway; however, this only happens if the current optimization level is set to `OptimizationLevel.O2`.

Reply:
OK, makes sense :+1: